netsim/providers/__init__.py

```diff
@@ -131,63 +138,104 @@ def create_extra_files_mappings(
           module=self.provider)
         continue
 
-      out_folder = f"{self.provider}_files/{node.name}"
-      bind_dict[f"{out_folder}/{file}"] = mapping         # note: node_files directory is flat
+      bind_dict[out_key] = mapping
 
     node[self.provider][outkey] = filemaps.dict_to_mapping(bind_dict)
 
   def create_extra_files(
       self,
       node: Box,
       topology: Box,
-      inkey: str = 'config_templates',
       outkey: str = 'binds') -> None:
 
     binds = node.get(f'{self.provider}.{outkey}',None)
     if not binds:
       return
 
     sys_folder = str(_files.get_moddir())+"/"
-    out_folder = f"{self.provider}_files/{node.name}"
 
     bind_dict = filemaps.mapping_to_dict(binds)
-    for file,mapping in bind_dict.items():
-      if not out_folder in file:                  # Skip files that are not mapped into the temporary provider folder
+    
+    node_data = {
+        **node.to_dict(),
+        'hostvars': topology.nodes.to_dict(),
+        'hosts': get_host_addresses(topology),
+        'addressing': topology.addressing.to_dict()
+    }
+
+    base_path = pathlib.Path(f"{self.provider}_files")
+
+    for file_str,mapping in bind_dict.items():
+      full_out_path = pathlib.Path(file_str)
+
+      # Only handle files placed under the provider's file directory
+      try:
+        full_out_path.relative_to(base_path)
+      except ValueError:
+        continue
+
+      # The template name is the file name part of the path
+      file_rel = full_out_path.name
+      if not file_rel:
+        # nothing to render
         continue
-      file_name = file.replace(out_folder+"/","")
-      template_name = self.find_extra_template(node,file_name,topology)
-      if template_name:
-        node_data = node + { 'hostvars': topology.nodes, 
-                             'hosts': get_host_addresses(topology),
-                             'addressing': topology.addressing }  # Needed for subnet prefix
-        if '/' in file_name:                      # Create subdirectory in out_folder if needed
-          pathlib.Path(f"{out_folder}/{os.path.dirname(file_name)}").mkdir(parents=True,exist_ok=True)
-        try:
-          templates.write_template(
-            in_folder=os.path.dirname(template_name),
-            j2=os.path.basename(template_name),
-            data=node_data.to_dict(),
-            out_folder=out_folder, filename=file_name)
-        except Exception as ex:
-          log.fatal(
-            text=f"Error rendering {template_name} into {file_name}\n{strings.extra_data_printout(str(ex))}",
-            module=self.provider)
 
+      # For shared files, extract the original file name (remove 'shared-' prefix)
+      template_fname = file_rel.removeprefix(self.SHARED_PREFIX)
+
+      template_name = self.find_extra_template(node, template_fname, topology)
+      if not template_name:
+        log.error(f"Cannot find template for {file_rel} on node {node.name}",log.MissingValue,'provider')
```

Comment:
We need to provide search path in this error message or it's too confusing to troubleshoot. Will open an issue for that.